Add Bash Completion Support

nickw444 opened the pull request:
I've implemented deep bash completion integration within the library. Work towards #75.  A bash completion script can call into any tool using Kingpin by using a new flag `--help-bash-completion` as the first argument, followed by subsequent user-entered arguments. 

Opening this PR to get some eyes on it, and see how it can be improved. Happy to add docs once we're happy with the overall structure of it (Since this does bring some new API):

- `type HintAction func(args []string) []string` A new type `HintAction`, a function that returns a list of options for completions.
- `func (a *ArgClause) HintAction(action HintAction) *ArgClause` to register a `HintAction` to call when completions are requested for an argument.
- `func (a *ArgClause) HintOptions(options ...string) *ArgClause` to register a static slice of strings to provide as completions when requested. Internally registers a `HintAction` that returns this slice.
- `func (a *FlagClause) HintAction(action HintAction) *FlagClause` - Same as described for `ArgClause`
- `func (a *FlagClause) HintOptions(options ...string) *FlagClause`  - Same as described for `ArgClause`

Additionally, FlagClause `EnumVar` will register all available `EnumVar` options as completion options too. 

Users will be presented with completion options for:
- sub-commands
- flags for a command (when `-` was the last user-entered argument)
- options for a flag (when the second last argument was a flag)

There's a bit of re-shuffling in this PR, including a new mixin (`cmdMixin`), due to some shared logic between the `Application` type and the `CmdClause` type for autocompletion

Additionally, a lot of logic from `Application.execute` has been moved to `Application.Parse` to allow different code paths for bash completion / normal execution.

An additional argument has been added to `Application.applyPreActions` to change whether or not sub commands should have preActions dispatched, since this causes issues with functionality such as `help`, and `--version`

Included are two scripts `support/bash_completion` and `support/zsh_completion` which should be modified and then sourced/installed by end users within their shells to enable bash completion.

alecthomas commented:
Hi @nickw444, thanks for the PR. I've had a look, but I'm still pondering some design/usability concerns. 

1. Foremost, I think, is that I *think* (I haven't stewed on this sufficiently to be sure) that I'd prefer if the entirety of the completion script were generated by Kingpin (eg. via a `--bash-completion-script` flag), rather than having a separately installed script. This is a usability concern. This would also allow the completion implementation to change.
2. I'm not 100% convinced that calling into the app to do completion is the best approach (vs. generating the full completion script up front).
3. It would be ideal if flags with values had `=` appended (eg. `--expires=` vs. `--expires `) and didn't add a space. I had a quick stab at this, but couldn't figure out how to tell bash not to add a space after completing.

I really like the hint approach, this could also be useful for eg. generating more useful help.

Anyway, I'll have a bit more of a think about it, but at the moment I'm thinking that if 1. above is addressed, I'll probably just merge it because it's a great start, then if the underlying implementation changes it will be easy to do.

alecthomas commented:
Also, I think the hint methods could use some documentation.

nickw444 commented:
Hey @alecthomas thanks for the comments. I understand your concerns in all of your above points.

**1.**

I've moved the "support" scripts into the `templates.go` file and implemented 2 new flags to generate these supporting files, `--completion-script-bash` and  `--completion-script-zsh`. However, my main concern is with package distribution for owners tools which use kingpin. 

Consider this: when distributing a package, (specifically a .deb), the package maintainer will usually add a `bash_completion` script to the package to be installed in `/etc/bash_completion.d`. Adding these flags does help, but it will need to be made clear bash completion doesn't come for free - If they want end users to get completions, they need to install these (generated) completion files in their package distributions. 

Alternatively, package maintainers need to inform users that they can add additional lines to their `.bash_profile` and `.zshrc` such as:

```
eval "$(my-cli-tool --completion-script-bash)"
```

However, personally, I feel this isn't a great end user experience. Anyway, with that all being said, it's definitely the right way to go to move the scripts into kingpin, where a package maintainer can generate these scripts after compilation and add them to the distribution.

**2.**

I originally felt that keeping completions outside of the app was the best idea, as it's more isolated and controlled (plus an easier feature IMO :P ), but, it didn't give users of kingpin great options for handling dynamic arguments without re-generating the script every time. 

A good example of such a use case is where you have a CLI tool which reads from an hosts file and provides completion of possible known hosts. By reaching into the application and performing a `HintAction` we can read new options on the fly without any re-generation. I think this will be made more clear with examples.

**3.**

I originally attempted this, and also implemented a hybrid, however due to how bash (and ZSH) handle completion selection, a space will always be added (unless there are multiple options to choose from). 

The original implementation I speak of would output every possible `--<flag>=<possible option for flag>` combination, but ended up being very messy in the case that the user was unsure initially of which flag to type (ie, just typing `--`).

**docs**

I've added documentation and tests. Let me know if you've got any additional comments :)

alecthomas commented:
The tests are failing, but once they're fixed up I'll take another look. Thanks!

alecthomas commented on app.go:
```diff
@@ -50,19 +48,23 @@ type Application struct {
 	a.HelpFlag.Bool()
 	a.Flag("help-long", "Generate long help.").Hidden().PreAction(a.generateLongHelp).Bool()
 	a.Flag("help-man", "Generate a man page.").Hidden().PreAction(a.generateManPage).Bool()
+	a.Flag("help-bash-completion", "Output possible completions for the given args.").Hidden().BoolVar(&a.completion)
```
I think I'd prefer this be named `---completion-bash`, as it has no real relationship to help.

alecthomas commented on app_test.go:
```diff
@@ -221,3 +223,127 @@ func TestDefaultEnvars(t *testing.T) {
+		assert.Equal(t, c.ExpectedOptions, args, "Expected != Actual: [%v] != [%v]. \nInput was: [%v]", c.ExpectedOptions, args, c.Args)
+	}
+
+}
```
Nice test, lovely.

alecthomas commented on completions.go:
```diff
@@ -0,0 +1,31 @@
+package kingpin
+
+type HintAction func(args []string) []string
```
Can you document this? What exactly is `args`?

alecthomas commented on examples/completion/main.go:
```diff
@@ -0,0 +1,96 @@
+	// You can provide hint options statically
+	nc.Flag("port", "Provide a port to connect to").
+		Required().
+		HintOptions("80", "443", "8080").
```
This is a good example of where `HintOptions()` could be useful. Maybe put that in the README as well?

alecthomas commented on examples/completion/main.go:
```diff
@@ -0,0 +1,96 @@
+	"fmt"
+	"os"
+
+	"gopkg.in/alecthomas/kingpin.v2"
```
This is the import causing the tests to fail.

alecthomas commented:
> I originally felt that keeping completions outside of the app was the best idea, as it's more isolated and controlled (plus an easier feature IMO :P ), but, it didn't give users of kingpin great options for handling dynamic arguments without re-generating the script every time.

That's a good point. Looking at the example completion code, it looks like it could be super powerful. Very nice.

> I originally attempted this, and also implemented a hybrid, however due to how bash (and ZSH) handle completion selection, a space will always be added (unless there are multiple options to choose from).

That's a bummer, but good to know you had the same thought :)

I think once you've addressed the few minor comments I made, this LGTM.

nickw444 commented:
I've made the requested modifications. I removed that `args` argument for `HintAction` since it was not used (in fact it was just left over from early development on this feature). 

However, I still can't get a green build. I've attempted using a relative package import, and not importing at all. What do you suggest I do to remediate this?

alecthomas commented:
The tests are failing because you're importing `gopkg.in/alecthomas/kingpin.v2` in the example `main.go`. You need to import `github.com/alecthomas/kingpin`.

alecthomas commented:
Apart from that, looks great, thanks!

alecthomas commented:
This is brilliant, thanks so much @nickw444. Much appreciated.

ches commented:
Very nice, thank you @nickw444!

nickw444 commented:
Hey @alecthomas just wondering when you'll be building a release with these changes for gopkg.in? Cheers

alecthomas commented:
I just pushed out v2.1.11

nickw444 commented:
Thanks!

alecthomas commented:
@nickw444 I've noticed that the command completion in particular seems inconsistent. _examples/curl doesn't seem to work at all. Any idea why? I haven't had time to look at it.

nickw444 commented:
I'll take a quick look now.

nickw444 commented:
I think I may have accidentally found the issue you were having:

1. Build _examples/curl. 
2. Run ./curl and perform tab completion.

```
./curl <TAB>
file:      ftp://     gopher://  http://    https://
```

The shell will respond with completion that doesn't seem consistent with kingpin completion. This is because we didn't source the kingpin completion script.

Run `eval "$(./curl --completion-script-zsh)"`. Try again:

```
./curl --<TAB>
--headers  --help     --timeout  --version

./curl <TAB>
(no output)
```

alecthomas commented:
I didn't see the first output, I saw no output, as you do with your last line. However, the curl example has commands and subcommands, which don't get completed. I'd expect:

```
./curl <TAB>
help get post
```

nickw444 commented:
I can reproduce this issue now. I'm sure this is because of the pattern being used to create this command line tool. Notice how the commands are instantiated from the kingpin library, rather than from an `app` struct:

```
...
get         = kingpin.Command("get", "GET a resource.").Default()
getFlag     = get.Flag("test", "Test flag").Bool()
getURL      = get.Command("url", "Retrieve a URL.").Default()
...
```

All the flags work because they're defined on the command.

Look at the completion example and how the commands are defined:

```
func addSubCommand(app *kingpin.Application, name string, description string) {
	c := app.Command(name, description).Action(func(c *kingpin.ParseContext) error {
		fmt.Printf("Would have run command %s.\n", name)
		return nil
	})
	c.Flag("nop-flag", "Example of a flag with no options").Bool()
}
...
func main() {
...
addSubCommand(app, "nmap", "Additional top level command to show command completion")
...
}
...
```

alecthomas commented:
They're functionally equivalent though, so I'm not sure why that would make a difference?

nickw444 commented:
Correct me if i'm wrong, but from what I can see is that they don't do the same thing:

`kingpin.Command`:
```
// global.go
var (
	// CommandLine is the default Kingpin parser.
	CommandLine = New(filepath.Base(os.Args[0]), "")
        ....
)

// Command adds a new command to the default parser.
func Command(name, help string) *CmdClause {
	return CommandLine.Command(name, help)
}
```

`kingpin.App.command`
```
// app.go:288
...
func (a *Application) Command(name, help string) *CmdClause {
	return a.addCommand(name, help)
}
...
```
```
// cmd.go:114
func (c *cmdGroup) addCommand(name, help string) *CmdClause {
	cmd := newCommand(c.app, name, help)
	c.commands[name] = cmd
	c.commandOrder = append(c.commandOrder, cmd)
	return cmd
}
```

When a command is added to an app, it is registered in the app's `commands` map. This map is used to generate completions. This is also the case for subcommands (since both `cmdClause` and `Application` inherit `cmdMixin`, which inherits `cmdGroup`)

nickw444 commented:
You're right about that. I'll take a look on Monday. Good find!

alecthomas commented:
👍
